arcs/src/components/spatial_entity.rs

```diff
@@ -0,0 +1,145 @@
+use crate::{
+    components::BoundingBox,
+    Vector,
+    primitives::{Arc},
+    algorithms::{Bounded},
+};
+use specs::{Entity, world::Index};
+use aabb_quadtree::{QuadTree, Spatial, ItemId};
+use quadtree_euclid::{TypedRect, TypedPoint2D, TypedSize2D};
+use std::collections::HashMap;
+
+pub(crate) type SpatialTree = QuadTree<SpatialEntity, f64, [(ItemId, TypedRect<f32, f64>); 0]>;
+
+/// A intermediate struct that maps an [`Entity`] to its [`BoundingBox`]
+/// 
+/// This is used to populate an efficient spatial lookup structure like a `QuadTree`
+#[derive(Debug)]
+pub struct SpatialEntity {
+    pub bounds: BoundingBox,
+    pub entity: Entity
+}
+
+impl Spatial<f64> for SpatialEntity {
+    fn aabb(&self) -> TypedRect<f32, f64> {
+        let bb = self.bounds;
+        TypedRect::<f32, f64>::new(
+            // TypedRects have their origin at the bottom left corner (this is undocumented!)
+            TypedPoint2D::new(bb.bottom_left().x as f32, bb.bottom_left().y as f32),
+            TypedSize2D::new(bb.width() as f32, bb.height() as f32))
+    }
+}
+
+impl SpatialEntity {
+    pub fn new(bounds: BoundingBox, entity: Entity) -> SpatialEntity {
+        SpatialEntity {
+            bounds,
+            entity
+        }
+    }
+}
+
+/// A global [`Resource`] for looking up which [`Entity`]s inhabit
+/// a given spatial point or region
+#[derive(Debug)]
+pub struct Space {
+    quadtree: SpatialTree,
+    ids: HashMap<Index, ItemId>
+}
+
+impl Default for Space {
+    fn default() -> Self {
+        Space {
+            quadtree: Self::default_tree(),
+            ids: HashMap::new()
+        }
+    }
+}
+
+impl Space {
+    // FIXME: Hard-code is bad-bad
+    const WORLD_RADIUS: f64 = 1_000_000.0;
+
+    // FIXME: We need to supply this in *DrawingUnits*
+    // as 1[meter] f.e. becomes meaningless when zoomed far in/out
+    const QUERY_POINT_RADIUS: f64 = 1.0;
+
+    fn default_tree() -> SpatialTree{
+        // Initialize quadtree
+        let size = BoundingBox::new(
+            Vector::new(-Self::WORLD_RADIUS, -Self::WORLD_RADIUS),
```

Comment:
As i commented [here](https://github.com/Michael-F-Bryan/arcs/issues/10#issuecomment-580292285), the trees `bounding_box` is hard-coded on initialization. If we try to insert something outside of those bounds afterwards it will panic.
I'm not a fan of how the `aabb_quadtree` handles this and I only see 3 options:
- Check all insertions in our API and rebuild the tree with a big enough bounding box every time *(seems really inefficient)*
- Allocate a **very** big world size *(also inefficient)*
- Use another `Quadtree`  implementation

Reply:
I did some more thinking on this:

- `Space` needs to implement `Default` so we can use it as a `Resource`, this inevitely means we have to define some default starting parameters, as we can not supply arguments to `Space::default()`.
- Memory-wise we should still be quite efficient, as very large empty space in a quad-tree should take the same amount of memory as small, filled spaces
- To be extra-safe I could write a wrapper around `aabb_quadtree.insert()` which detects objects outside the bounds of the *tree-root* and resize the tree if needed
- Resizing basically means re-building the whole tree again, but with a large enough *BoundingBox* so the newly added `SpatialEntity` fits in

Reply:
> Space needs to implement Default so we can use it as a Resource, this inevitely means we have to define some default starting parameters, as we can not supply arguments to Space::default().

Could we add `Space` as a resource in the bookkeeping system's `setup()` method then use `ReadExpect` when trying to access it?

> Resizing basically means re-building the whole tree again, but with a large enough BoundingBox so the newly added SpatialEntity fits in

That's annoying. 

I was kinda hoping you could implement a resize operation by taking the quad tree's "head" node and wrapping it in a larger node, almost like the tree equivalent of adding an item to the front of a linked list. You'd be left with a really unbalanced tree (the entire previous tree is in one quadrant of the tree's head node), but it turns a resize into an `O(1)` operation where we just copy some pointers around.

Reply:
> Could we add Space as a resource in the bookkeeping system's setup() method then use ReadExpect when trying to access it?

This seems like an unnecessary mix of two unrelated systems. So I'd rather stick to `Space` being it's own system

> almost like the tree equivalent of adding an item to the front of a linked list.
There is nothing in the API that woudl allow for this.

It really looks like we need a better backend for our `Space`